### PR TITLE
Fix asyncio actor race condition

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

While modifying `test_dynres.py` to use an asyncio actor as a signal for timing instead of random object IDs, I discovered a bug where asyncio actor tasks could get placed in the scheduling queue before the is_asyncio flag was set by the creation task, causing them to block the actor instead of yielding the event loop. This patch both fixes the bug by delaying the is_async check until after the creation task runs and adds the changes to `test_dynres.py`.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 1]
```diff
@@ -178,44 +178,17 @@ void CoreWorkerDirectTaskReceiver::Init(rpc::ClientFactoryFn client_factory,
...
-    is_asyncio_ = true;
-  }
-};
-
 void CoreWorkerDirectTaskReceiver::HandlePushTask(
     const rpc::PushTaskRequest &request, rpc::PushTaskReply *reply,
     rpc::SendReplyCallback send_reply_callback) {
   RAY_CHECK(waiter_ != nullptr) << "Must call init() prior to use";
   const TaskSpecification task_spec(request.task_spec());
-  RAY_LOG(DEBUG) << "Received task " << task_spec.DebugString();
```
Comment (simon-mo) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: revert this line?

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 2]
```diff
@@ -339,6 +356,8 @@ class SchedulingQueue {
     }
   }
 
+  // Worker context.
```
Comment (simon-mo) [inline src/ray/core_worker/transport/direct_actor_transport.h]: this is owned by CoreWorkerDirectTaskReceiver right?

**Review by simon-mo — APPROVED** [review]

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 2 above)
Comment (edoakes) [reply src/ray/core_worker/transport/direct_actor_transport.h]: owned by the core worker actually

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 3]
```diff
@@ -283,6 +279,24 @@ class SchedulingQueue {
...
+      if (!is_asyncio_) {
+        RAY_LOG(DEBUG) << "Setting direct actor as async, creating new fiber thread.";
+        fiber_state_.reset(new FiberState(max_concurrency));
+        is_asyncio_ = true;
+      }
+    } else {
+      // If this is a concurrency actor (not async), initialize the thread pool once.
+      if (max_concurrency != 1 && !pool_) {
+        RAY_LOG(INFO) << "Creating new thread pool of size " << max_concurrency;
+        pool_.reset(new BoundedExecutor(max_concurrency));
```
Comment (raulchen) [inline src/ray/core_worker/transport/direct_actor_transport.h:296]: @edoakes This PR changes the behavior of (non-asyncio) concurrent actor calls. Previously, there will be only one thread pool of size N. Now, we will create a new thread pool of size N for each caller. This leads to creating too many threads. 
I read your PR description. It seems that this change is unintentional?

Comment (raulchen) [reply src/ray/core_worker/transport/direct_actor_transport.h:296]: I think we should still put the pool and the fiber state in CoreWorkerDirectTaskReceiver. We can address the original issue by only calling SetMaxActorConcurrency and SetActorAsAsync in actor creation tasks.

Comment (raulchen) [reply src/ray/core_worker/transport/direct_actor_transport.h:296]: @edoakes Each `SchedulingQueue` has a thread pool. And in `CoreWorkerDirectTaskReceiver`, we will create a new `SchedulingQueue` for each caller. 

```python
  /// Queue of pending requests per actor handle.
  /// TODO(ekl) GC these queues once the handle is no longer active.
  std::unordered_map<TaskID, std::unique_ptr<SchedulingQueue>> scheduling_queue_;
```

```python
  auto it = scheduling_queue_.find(task_spec.CallerId());
  if (it == scheduling_queue_.end()) {
    auto result = scheduling_queue_.emplace(
        task_spec.CallerId(), std::unique_ptr<SchedulingQueue>(new SchedulingQueue(
                                  task_main_io_service_, *waiter_, worker_context_)));
    it = result.first;
  }
```

Comment (raulchen) [reply src/ray/core_worker/transport/direct_actor_transport.h:296]: Filed an issue https://github.com/ray-project/ray/issues/14894
will fix it shortly.

Comment (edoakes) [reply src/ray/core_worker/transport/direct_actor_transport.h:296]: Ah yeah I see your point; good find, we should definitely only be creating this many threads per actor

**Pull request merged**